core/src/transport.rs

```diff
@@ -134,6 +134,14 @@ pub trait Transport {
         addr: Multiaddr,
     ) -> Result<Self::Dial, TransportError<Self::Error>>;
 
+    ///
+    fn dial_with_new_port(
+        &mut self,
+        _addr: Multiaddr,
+    ) -> Result<Self::Dial, TransportError<Self::Error>> {
+        unimplemented!()
+    }
```

Comment:
Not a fan of this API at all but I also don't have a better idea in the short-term. Long-term, I'd like to get rid of the `Dial` associated type (see https://github.com/libp2p/rust-libp2p/pull/3436).